File: app/models/location.rb
# frozen_string_literal: true

# == Schema Information
#
# Table name: locations
#
#  id           :uuid             not null, primary key
#  properties   :jsonb            not null
#  slug         :string
#  created_at   :datetime         not null
#  updated_at   :datetime         not null
#  name         :string           not null
#  country_code :string           not null
#  address      :string
#  zip          :string
#  city         :string
#  lat          :decimal(, )
#  lon          :decimal(, )
#  category     :string
#  timezone     :string
#  bluf_id      :string
#

class Location < ApplicationRecord
  include DocumentSerializable
  extend FriendlyId
  friendly_id :name, use: :slugged

  has_many :events, dependent: :restrict_with_exception
  has_and_belongs_to_many :owners, class_name: "User"

  validates :name, :country_code, presence: true

  def country
    Country[country_code.to_sym]
  end

  def country_name
    country&.name
  end

  def description
    [name, city, country.name].compact.uniq.join(", ")
  end

  def coordinates?
    lat.present? && lon.present?
  end
end

File: src/models/location.ts
import { Syncable } from './syncable';
import { t } from '@nerdgeschoss/i18n';

export interface Location extends Syncable {
  id: string;
  slug: string;
  name: string;
  abstract?: string;
  description?: string;
  city?: string;
  address?: string;
  lat?: number;
  lon?: number;
  zip?: string;
  countryCode: string;
  countryName?: string;
  category: 'city' | 'bar' | 'region' | 'shop' | 'hotel' | 'venue';
  timezone?: string;
}

export function countryName(code: string): string {
  return (t('location.countries') as any)[code];
}

export function locationDescription(location: {
  name: string;
  city?: string | null;
  countryName?: string | null;
  countryCode?: string;
}): string {
  const line = [location.name];
  if (location.city && location.name !== location.city) {
    line.push(location.city);
  }
  line.push(location.countryName || countryName(location.countryCode!));
  return line.join(', ');
}

export function isVenue(location: { category: string | null }): boolean {
  return location.category !== 'venue';
}

export function extractCoordinates(location: {
  lat?: number | null;
  lon?: number | null;
}): { lat: number; lon: number } | undefined {
  if (!location.lat || !location.lon) {
    return;
  }
  return { lat: location.lat, lon: location.lon };
}

export function matchesTerm(
  location: Partial<Location>,
  term: string
): boolean {
  const locationName =
    (location.name || '').toLocaleLowerCase() +
    (location.address || '').toLocaleLowerCase() +
    (location.countryName || '').toLocaleLowerCase() +
    (location.countryCode &&
      (countryName(location.countryCode) || '').toLocaleLowerCase());
  const normalizedTerm = term.toLocaleLowerCase();
  return locationName.includes(normalizedTerm);
}
